### PR TITLE
Add Will It Shuffle Sample

### PR DESCRIPTION
**Pull request by jcorbin**
Submitting my "hello world" program, an adaptation of Mike Bostock's [Will It Shuffle](https://bost.ocks.org/mike/shuffle/compare.html) comparison from the D3.js data viz world.

I used this program to both orient on the language, and to validate #58 .

Perhaps at least useful as a conversation piece, even if it's not a useful sample to merge since, at least for me, it points to potential language/stdlib improvements:
- clamp would be nice to use for numeric structs
- I've not fully climbed "Spline Maths Mountain", so I don't know yet if there's an easy answer to my Basis v Cardinal spline question
- exhibits my "singleton coroutine" situation, where in I wish I could write `let mycoro = coroutine def(): ...` to have an "anonymous singleton coroutine" (for shuffler and uier)
- a note about a missing strings stdlib, which I implement a single typical (join) function

### CODE REVIEW

File: samples/will_it_shuffle.lobster [hunk 1]
```diff
@@ -0,0 +1,228 @@
+// adaptation of https://bost.ocks.org/mike/shuffle/compare.html
+//
```
Comment (aardappel) [inline samples/will_it_shuffle.lobster]: You can add yourself as the author, seeing as its one of the few files not by me.

File: samples/will_it_shuffle.lobster [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+// TODO
+// - vectorize the b-spline
+// - more shufflers
+// - measure reshuffle time
+// - swappable RNG
```
Comment (aardappel) [inline samples/will_it_shuffle.lobster]: Feel free to add this to TODO.txt, or add a built-in function `set_rnd_generator` or whatever.

File: samples/will_it_shuffle.lobster [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        let j = rnd(xs.length)
+        xs[i] = xs[j]
+        xs[j] = x
+
+def push_remove(xs):
+    // std.lobster original
+    for(xs.length) i:
+        xs.push(xs.remove(rnd(xs.length - i)))
+
+def fisher_yates(xs):
```
Comment (aardappel) [inline samples/will_it_shuffle.lobster]: use vs in `std` if that gets merged

File: samples/will_it_shuffle.lobster [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    div( (_ >> 8)  & 0xFF, 255),
+    div(  _        & 0xFF, 255),
+    1,
+}
+
+def spline_value(values, t):
+    // uniforem b-spline interpolation adapted from d3-scale-chromactic
+
+    // TODO can we translate from this Basis Spline (B-spline) to a cardinal
+    // spline? Or should lobster also provide a builtin for B-splines?
```
Comment (aardappel) [inline samples/will_it_shuffle.lobster]: Yup, feel free to add other spline functions, similar to the code for cardinal spline. That is nice to have, and generally good to have very re-usable core code in C++.

File: samples/will_it_shuffle.lobster [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+    t = clamp(t, 0, 1)
+    let n = values.length - 1
+    let i = clamp(floor(t * n), 0, n-1)
```
Comment (aardappel) [inline samples/will_it_shuffle.lobster]: And yes, if there's no vector clamp, there should be. Vector all the things!

File: samples/will_it_shuffle.lobster [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+        if shuffling or changed or last_mode != mode:
+            reshuffle()
+            last_mode = mode
+        next(nil)
+    return nil
+let shuffler = coroutine shuffle_routine()
+
+def viz():
+    let status = join(" ", [
```
Comment (aardappel) [inline samples/will_it_shuffle.lobster]: Doesn't this do pretty much exactly what `join` does?

https://github.com/aardappel/lobster/blob/dffd8a8944cbe5d81789b22926f725b013183cd7/dev/src/builtins.cpp#L504-L515

Comment (jcorbin) [reply samples/will_it_shuffle.lobster]: Oops, had missed that one.

File: samples/will_it_shuffle.lobster (hunk 2 above)
Comment (jcorbin) [reply samples/will_it_shuffle.lobster]: Lofted most out into `dev/TODO.txt`, and expanded on my own internal musings on where the PRNG stdlib should head...

File: samples/will_it_shuffle.lobster (hunk 3 above)
Comment (jcorbin) [reply samples/will_it_shuffle.lobster]: Simplified to match, but leaving both dual implemented here for clarity of the demo fwiw

File: samples/will_it_shuffle.lobster (hunk 4 above)
Comment (jcorbin) [reply samples/will_it_shuffle.lobster]: Neat, will look into doing so after this PR; I think I'm going to actually try writing something more like a game first fwiw :-)

File: samples/will_it_shuffle.lobster (hunk 5 above)
Comment (jcorbin) [reply samples/will_it_shuffle.lobster]: Yup, left it as a TODO for now, rather than keep sprawling out in this PR :-)

File: dev/TODO.txt [hunk 7]
```diff
@@ -1,3 +1,15 @@
+- improved random number generation:
+  - swappable generators such as xorshift, pcg32/64, and even just a generic lcg
+  - a good lobster choice may be to abstract them as coroutines, or at least a
+    higher order `def generator(consume):` protocol
```
Comment (aardappel) [inline dev/TODO.txt]: there's currently no way to create a "builtin co-routine" (one written in C++). I'd suggest keeping this simple: just a function that can select which type of rng is currently active.

**Comment by aardappel** [comment]
Thanks! Future improvements can be in followups.

**Pull request merged**